src/main/java/com/google/sps/servlets/ListTypeRecipesServlet.java

```diff
@@ -44,19 +45,35 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     
     List<Recipe> recipes = new ArrayList<>();
+    List<Key> deletedKeys = new ArrayList<>();
     for (Key key : keys) {
       try {
-        // TODO: do not include private recipes
         Entity entity = datastore.get(key);
         long id = key.getId();
         String name = (String) entity.getProperty("name");
         String description = (String) entity.getProperty("description");
         String blobkey = (String) entity.getProperty("imageBlobKey");
-
-        Recipe recipe = new Recipe(id, name, blobkey, description);
-        recipes.add(recipe);
+        boolean published = (boolean) entity.getProperty("published");
+        String authorID = (String) entity.getProperty("authorID");
+        // will add the recipe if the recipe belongs to the user or if it's published
+        if (published || loggedInAuthorID.equals(authorID)) {
+          Recipe recipe = new Recipe(id, name, blobkey, description);
+          recipes.add(recipe);
+        }
       } catch (EntityNotFoundException e) {
-        System.out.println("ListTypeRecipesServlet: Entity not found with key from user key list. This can happen when the recipe is deleted.");
+        // this happens when a recipe was deleted.
+        deletedKeys.add(key);
+      }
+    }
+
+    for (Key key: deletedKeys) {
+      // TODO: inform user how many recipes were deleted
+      if (type.equals("cookbook")) {
+        user.removeCookbookKey(key);
+      } else if (type.equals("planner")) {
+        user.removePlannerKey(key);
+      } else {
+         System.out.println("ListTypeRecipesServlet: invalid type");
```

Comment:
Instead of iterating through the deleted keys and maintaining a separate array list, can you move this logic into the catch block?

Also, can you add the type to the error message that's printed? Like you do in ManageListServlet.

Reply:
like delete the key in the catch block? I tried to do that and it threw an error

Reply:
will do the second!